### ISSUE TITLE
`MacroIf` then/else missing location information

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#14884]
I was looking into #14880 a bit and noticed that the `then` and `else` nodes of `MacroIf` do not have location information. E.g.

```cr
macro test(val)
  {% pp "#{val.cond} => #{val.cond.line_number} => #{val.cond.end_line_number}" %}
  {% pp "#{val.then} => #{val.then.line_number} => #{val.then.end_line_number}" %}
  {% pp "#{val.else} => #{val.else.line_number} => #{val.else.end_line_number}" %}
end

test({% if 1 == val %}
  pp "one!"
  pp "bar"
{% else %}
  pp "not one"
  pp "bar"
{% end %})
```

Currently yields:
```cr
"1 == val => 7 => 7"
"\n  pp \"one!\"\n  pp \"bar\"\n => nil => nil"
"\n  pp \"not one\"\n  pp \"bar\"\n => nil => nil"
```

For this specific example I was able to go to: https://github.com/crystal-lang/crystal/blob/b9ab9968b85ce2b99a875abd4360b1a432e27a98/src/compiler/crystal/syntax/parser.cr#L3214 and add a `.at @token.location`. which made the output now yield:
```cr
"1 == val => 7 => 7"
"\n  pp \"one!\"\n  pp \"bar\"\n => 7 => nil"
"\n  pp \"not one\"\n  pp \"bar\"\n => 10 => nil"
```
Which is better, but I would have expected the start lines to be `7, 8, 11`. I'm thinking that first newline when printing the node is making it technically start on the previous line. Not sure if that is expected or not? Or how to best proceed here.

**Comment by HertzDevil** [comment]
The `then` and `else` clauses here are both `MacroLiteral` nodes, whose leading and trailing whitespace is significant, so yes, 7 and 10 are the expected behavior (and the end lines would be 10 and 13 respectively)

**Comment by Blacksmoke16** [comment]
Any good ideas on how to work around the fact `then` and `else` nodes are technically on the same line as the condition? 

E.g. when parsing
```cr
{% begin %}
  {% v = 6 %}
  {% if v == 2 %}
    {{v * 2}}
  {% elsif v == 3 %}
    {{v * 5}}
  {% else %}
    {{v}}
  {% end %}
{% end %}
```
The nodes come in as:
```
{"v == 2", Crystal::Call, /home/george/dev/git/crystal/cov.cr:13:9, false}
{"2", Crystal::NumberLiteral, /home/george/dev/git/crystal/cov.cr:13:14, false}
{"    {{ v * 2 }}  ", Crystal::Expressions, /home/george/dev/git/crystal/cov.cr:13:18, true}
```
It'll first parse the condition, which is `v == 2` and `2` lines here. Since in this case, the conditional evaluates to falsey, I'm marking the `node.then` as missed, hence the `true` as the last value in the tuple. But because its location is the same as that of the condition, it messes things up by making my logic think that line 3 has no hits, when in reality line 3 should be hit once and line 4 zero times.

Same thing happens for the `elsif`:

```cr
{"v == 3", Crystal::Call, /home/george/dev/git/crystal/cov.cr:5:12, false}
{"3", Crystal::NumberLiteral, /home/george/dev/git/crystal/cov.cr:5:17, false}
{"    {{ v * 5 }}  ", Crystal::Expressions, /home/george/dev/git/crystal/cov.cr:5:21, true}
```

Maybe something on the lines of creating a new location for the if/else nodes that account for the number of newlines in the preceding `MacroLiteral`?  :shrug:

**Comment by straight-shoota** [comment]
Not sure I understand what you mean. But I assume it's in the context of macro coverage?
How does other code coverage work with conditionals and their conditions?

**Comment by Blacksmoke16** [comment]
Sorry, yea I should have clarified. From what I can tell other tools do a similar thing, e.g. https://github.com/simplecov-ruby/simplecov?tab=readme-ov-file#branch-coverage-ruby--25. 

```cr
macro i(node)
  {%
    pp node.line_number      # => 8
    pp node.then.line_number # => 8
  %}
end

i({% if v == 2 %}
    {{v * 2}}
  {% end %})
```
Like it makes sense to me _why_ this is the case, but would def be a lot easier for code coverage if `node.then.line_number # => 9` given that's the line that actually wasn't hit. I was able to seemingly find a workaround so I just went with that for now :shrug:.
